autorest/validations.go

```diff
@@ -0,0 +1,359 @@
+		}
+	case minItems:
+		if _, ok := v.Rule.(int); !ok {
+			return createError(x, v, fmt.Sprintf("rule must be integer for %v constraint; got: %v", v.Name, v.Rule))
+		}
+		if z.Len() < v.Rule.(int) {
+			return createError(x, v, fmt.Sprintf("mininum item limit is %v; got: %v", v.Rule, z.Len()))
+		}
+	case uniqueItems:
+		if z.Kind() == reflect.Array || z.Kind() == reflect.Slice {
+			if !checkForUniqueInArray(x) {
+				return createError(x, v, fmt.Sprintf("all items in parameter %v must be unique; got:%v", v.Target, z))
+			}
+		} else if z.Kind() == reflect.Map {
+			if !checkForUniqueInMap(x) {
+				return createError(x, v, fmt.Sprintf("all items in parameter %v must be unique; got:%v", v.Target, z))
+			}
+		} else {
+			return createError(x, v, fmt.Sprintf("type must be array, slice or map for constraint %v; got: %v", v.Name, z.Kind()))
+		}
+	default:
+		return createError(x, v, fmt.Sprintf("constraint %v is not applicable to array, slice and map type", v.Name))
+	}
+
+	if v.Chain != nil {
+		return Validate([]Validation{
+			{
+				TargetValue: x,
+				Constraints: *v.Chain,
+			},
+		})
+	}
+	return nil
+}
+
+func checkNil(x interface{}, v Constraint) error {
+	if _, ok := v.Rule.(bool); !ok {
+		return createError(x, v, fmt.Sprintf("rule must be bool value for %v constraint; got: %v", v.Name, v.Rule))
+	}
+	if v.Rule.(bool) {
+		return createError(x, v, "value can not be null; required parameter")
+	}
+	return nil
+}
+
+func checkEmpty(x interface{}, v Constraint) error {
+	if _, ok := v.Rule.(bool); !ok {
+		return createError(x, v, fmt.Sprintf("rule must be bool value for %v constraint; got: %v", v.Name, v.Rule))
+	}
+
+	if v.Rule.(bool) {
+		return createError(x, v, "value can not be null or empty; required parameter")
+	}
+	return nil
+}
+
+func checkForUniqueInArray(x interface{}) bool {
+	z := reflect.ValueOf(x)
+	if x == nil || z.Len() == 0 {
+		return false
+	}
+	arrOfInterface := make([]interface{}, z.Len())
+
+	for i := 0; i < z.Len(); i++ {
+		arrOfInterface[i] = z.Index(i).Interface()
+	}
+
+	m := make(map[interface{}]bool)
+	for _, val := range arrOfInterface {
+		if m[val] {
+			return false
+		}
+		m[val] = true
+	}
+	return true
+}
+
+func checkForUniqueInMap(x interface{}) bool {
+	z := reflect.ValueOf(x)
+	if x == nil || z.Len() == 0 {
+		return false
+	}
+	mapOfInterface := make(map[interface{}]interface{}, z.Len())
+
+	keys := z.MapKeys()
+	for _, k := range keys {
+		mapOfInterface[k.Interface()] = z.MapIndex(k).Interface()
+	}
+
+	m := make(map[interface{}]bool)
+	for _, val := range mapOfInterface {
+		if m[val] {
+			return false
+		}
+		m[val] = true
+	}
+	return true
+}
+
+// ValidationError stroes detailed validation error
+type ValidationError struct {
+	Constraint  string
+	Target      string
+	TargetValue interface{}
+	Details     string
+}
+
+func (e ValidationError) Error() string {
+	return fmt.Sprintf("autorest/azure: validation error: Parameter=%v Value=%v Constraint=%v Details=%v",
```

Comment:
also if e.TargetValue is map or an array, this message is going to be quite unreadable.

so perhaps move Constraint before Value?